src/main/kotlin/org/rust/cargo/project/workspace/CargoWorkspace.kt

```diff
@@ -21,6 +24,7 @@ import java.util.*
  * [org.rust.cargo.project.model.CargoProjectsService] manages workspaces.
  */
 interface CargoWorkspace {
+    val project: Project
```

Comment:
I feel very uneasy about addition of this field: it introduces a tight coupling between the two, previously completely independent, things. It's a useful property that `CargoWorkspace` knows nothing about IDEA's model. 

Instead of adding fields to this data structure, I suggest storing the caches in a hash-map keyed by packages. The hash map can live on some UserDataHolder or it can be encapsulated as a service (or component? That XML-thing with a lifecycle.)

Reply:
More specifically, `CargoProject` is the thing which probably should store this data. See, for example, 

https://github.com/intellij-rust/intellij-rust/blob/fc6c52b2a8b896326426983370b070e9babf99af/src/main/kotlin/org/rust/lang/core/resolve/KnownItems.kt#L28-L32